README.md
# flureejs-request


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
(modification: no type change headlines) and this project adheres to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1.3] - 2020-09-07

### Changed

- Facilities for getting msg, digest and signature
- Documentation and test improvements

## [0.1.2] - 2020-08-30

### Changed

- Working Karma into Chrome and Firefox
- Documentation and test improvements

## [0.1.1] - 2020-08-23

### Changed

- Changes for use flureejs-config package
- Documentation and test improvements

## [0.1.0] - 2020-08-21

- Basic functionality

[0.1.3]: https://github.com/StylusFrost/flureejs-request/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/StylusFrost/flureejs-request/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/StylusFrost/flureejs-request/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/StylusFrost/flureejs-request/releases/v0.1.0


test/types.ts
// The type of ./requets.json
export type RequestsJsonEntry = {
  privateKey: string
  sendersAuthID: string
  type: string
  raw: string[]
}


test/index.spec.ts
/* tslint:disable no-invalid-this */
import * as assert from 'assert'

import Request from '../src'
import { RequestsJsonEntry } from './types'

import { zeros, privateToPublic, toBuffer } from 'flureejs-utils'

const requestFixtures: RequestsJsonEntry[] = require('./requests.json')

describe('[Request]: Basic functions', function() {
  const requests: Request[] = []
  it('should decode requests', function() {
    requestFixtures.slice(0, 3).forEach(function(request: any) {
      const pQ = new Request(request.raw)
      assert.equal('0x' + pQ.param.toString('hex'), request.raw[0])
      assert.equal('0x' + pQ.type.toString('hex'), request.raw[1])
      assert.equal('0x' + pQ.host.toString('hex'), request.raw[2])
      assert.equal('0x' + pQ.db.toString('hex'), request.raw[3])
      assert.equal('0x' + pQ.formattedDate.toString('hex'), request.raw[4])
      assert.equal('0x' + pQ.v.toString('hex'), request.raw[6])
      assert.equal('0x' + pQ.r.toString('hex'), request.raw[7])
      assert.equal('0x' + pQ.s.toString('hex'), request.raw[8])
      requests.push(pQ)
    })
  })

  it('should verify Signatures', function() {
    requests.forEach(function(request) {
      assert.equal(request.verifySignature(), true)
    })
  })

  it('should not verify Signatures', function() {
    requests.forEach(function(request) {
      request.s = zeros(32)
      assert.equal(request.verifySignature(), false)
    })
  })

  it('should give a string about not verifing Signatures', function() {
    requests.forEach(function(request) {
      assert.equal(request.validate(true).slice(0, 17), 'Invalid Signature')
    })
  })

  it('should validate', function() {
    requests.forEach(function(request) {
      assert.equal(request.validate(), false)
    })
  })

  it('should sign request', function() {
    requests.forEach(function(request, i) {
      if (requestFixtures[i].privateKey) {
        const privKey = new Buffer(requestFixtures[i].privateKey, 'hex')
        request.sign(privKey)
      }
    })
  })
  it("should get sender's AuthID after signing it", function() {
    requests.forEach(function(request, i) {
      if (requestFixtures[i].privateKey) {
        assert.equal(request.getSenderAuthID(), requestFixtures[i].sendersAuthID)
      }
    })
  })

  it("should get sender's public key after signing it", function() {
    requests.forEach(function(request, i) {
      if (requestFixtures[i].privateKey) {
        assert.equal(
          request.getSenderPublicKey().toString('hex'),
          privateToPublic(new Buffer(requestFixtures[i].privateKey, 'hex')).toString('hex'),
        )
      }
    })
  })

  it("should get sender's authID after signing it (second call should be cached)", function() {
    requests.forEach(function(request, i) {
      if (requestFixtures[i].privateKey) {
        assert.equal(request.getSenderAuthID(), requestFixtures[i].sendersAuthID)
        assert.equal(request.getSenderAuthID(), requestFixtures[i].sendersAuthID)
      }
    })
  })

  it('should verify signing it', function() {
    requests.forEach(function(request, i) {
      if (requestFixtures[i].privateKey) {
        assert.equal(request.verifySignature(), true)
      }
    })
  })

  it('should accept lesser r values', function() {
    const request = new Request()
    request.r = toBuffer('0x0005')
    assert.equal(request.r.toString('hex'), '05')
  })
})


src/types.ts
/**
 * Any object that can be transformed into a `Buffer`
 */
export interface TransformableToBuffer {
  toBuffer(): Buffer
}

/**
 * A hex string prefixed with `0x`.
 */
export type PrefixedHexString = string

/**
 * A Buffer, hex string prefixed with `0x`, Number, or an object with a toBuffer method such as BN.
 */
export type BufferLike = Buffer | TransformableToBuffer | PrefixedHexString | number

export interface RequestData {
  /**
   * This will contain the data of the request
   */
  param?: BufferLike

  /**
   * The request's type.
   */
  type?: BufferLike

  /**
   * The request's host to.
   */
  host?: BufferLike

  /**
   * The request's db to.
   */
  db?: BufferLike

  /**
   * The request's date.
   */
  formattedDate?: BufferLike

  /**
   * If an authority is signing
   */
  auth?: BufferLike

  /**
   * EC recovery ID.
   */
  v?: BufferLike

  /**
   * EC signature parameter.
   */
  r?: BufferLike

  /**
   * EC signature parameter.
   */
  s?: BufferLike
}


src/index.ts
import {
  ecsign,
  publicToAuthID,
  ecrecover,
  toBuffer,
  sha256,
  defineProperties,
  bufferToInt,
} from 'flureejs-utils'
import { RequestData } from './types'

export default class Request {
  public raw!: Buffer[]
  public param!: Buffer
  public type!: Buffer
  public host!: Buffer
  public db!: Buffer
  public formattedDate!: Buffer
  public auth!: Buffer
  public v!: Buffer
  public r!: Buffer
  public s!: Buffer

  protected _from?: Buffer
  private _senderPubKey?: Buffer

  /**
   * Creates a new request from an object with its fields' values.
   *
   * @param data - A request can be initialized with object containing them by name.
   *
   * @example
   * ```js
   * const RequestData = {
   *   param: '0x0000000000000000000000',
   *   type: '0x09184e72a000',
   *   host: '0x2710',
   *   db: '0x0000000000000000000000000000000000000000',
   *   formattedDate:'0x0000000000000000',
   *   auth: '0x00',
   *   v: '0x1c',
   *   r: '0x5e1d3a76fbf824220eafc8c79ad578ad2b67d01b0c2425eb1f1347e8f50882ab',
   *   s: '0x5bd428537f05f9830e93792f90ea6a3e2d1ee84952dd96edbae9f658f831ab13'
   * };
   * const q = new Request(RequestData);
   * ```
   */

  constructor(data: RequestData = {}) {
    // Define Properties
    const fields = [
      {
        name: 'param',
        allowZero: false,
        default: new Buffer([]),
      },
      {
        name: 'type',
        allowZero: false,
        default: new Buffer([]),
      },
      {
        name: 'host',
        allowZero: false,
        default: new Buffer([]),
      },
      {
        name: 'db',
        allowZero: false,
        default: new Buffer([]),
      },
      {
        name: 'formattedDate',
        allowZero: false,
        default: new Buffer([]),
      },
      {
        name: 'auth',
        allowZero: true,
        default: new Buffer([]),
      },
      {
        name: 'v',
        allowZero: true,
        default: new Buffer([]),
      },
      {
        name: 'r',
        length: 32,
        allowZero: true,
        allowLess: true,
        default: new Buffer([]),
      },
      {
        name: 's',
        length: 32,
        allowZero: true,
        allowLess: true,
        default: new Buffer([]),
      },
    ]

    // attached serialize
    defineProperties(this, fields, data)

    /**
     * @property {Buffer} from (read only) sender address of this transaction, mathematically derived from other parameters.
     * @name from
     * @memberof Transaction
     */
    Object.defineProperty(this, 'from', {
      enumerable: true,
      configurable: true,
      get: this.getSenderAuthID.bind(this),
    })

    this._validateV(this.v)
    this._overrideVSetterWithValidation()
  }

  private _validateV(v?: Buffer): void {
    if (v === undefined || v.length === 0) {
      return
    }

    const vInt = bufferToInt(v)

    if (vInt === 27 || vInt === 28) {
      return
    }
  }

  private _overrideVSetterWithValidation() {
    const vDescriptor = Object.getOwnPropertyDescriptor(this, 'v')!

    Object.defineProperty(this, 'v', {
      ...vDescriptor,
      set: v => {
        if (v !== undefined) {
          this._validateV(toBuffer(v))
        }
        vDescriptor.set!(v)
      },
    })
  }

  /**
   * returns the sender's authID
   */
  getSenderAuthID(): Buffer {
    if (this._from) {
      return this._from
    }
    const pubkey = this.getSenderPublicKey()
    this._from = publicToAuthID(pubkey)
    return this._from
  }

  /**
   * returns the public key of the sender
   */
  getSenderPublicKey(): Buffer {
    if (!this.verifySignature()) {
      throw new Error('Invalid Signature')
    }

    // If the signature was verified successfully the _senderPubKey field is defined
    return this._senderPubKey!
  }

  /**
   * Determines if the signature is valid
   */
  verifySignature(): boolean {
    try {
      const v = bufferToInt(this.v)
      this._senderPubKey = ecrecover(this.msg(), v, this.r, this.s)
    } catch (e) {
      return false
    }
    return !!this._senderPubKey
  }

  signature(): Buffer {
    const newR = this.r[0] & 0x80 ? Buffer.concat([hexToUnit8Array('00'), this.r]) : this.r
    const newS = this.s[0] & 0x80 ? Buffer.concat([hexToUnit8Array('00'), this.s]) : this.s
    const result =
      '02' +
      newR.length.toString(16) +
      newR.toString('hex') +
      '02' +
      newS.length.toString(16) +
      newS.toString('hex')

    const signature = Buffer.from(
      this.v.toString('hex') + '30' + Buffer.from(result, 'hex').length.toString(16) + result,
      'hex',
    )

    const keyID = Buffer.compare(this.auth, Buffer.alloc(0)) != 0 ? this.auth.toString() : 'na'

    return Buffer.from(
      'keyId="' +
        keyID +
        '",headers="(request-target) x-fluree-date digest",algorithm="ecdsa-sha256",signature="' +
        signature.toString('hex') +
        '"',
    )

    function hexToUnit8Array(str: string) {
      return new Uint8Array(Buffer.from(str, 'hex'))
    }
  }

  digest(): Buffer {
    const digest = sha256(Buffer.from(this.param.toString('hex').normalize(), 'hex'))
    return Buffer.from(`SHA-256=${digest.toString('base64')}`)
  }

  private msg(): Buffer {
    const uri = '/fdb/' + this.db.toString() + '/' + this.type.toString()
    return Buffer.from(`(request-target): post ${uri}
    x-fluree-date: ${this.formattedDate.toString()}
    digest: ${this.digest().toString()}`)
  }

  /**
   * Validates the signature and checks to see if it has enough gas.
   */
  validate(): boolean
  validate(stringError: false): boolean
  validate(stringError: true): string
  validate(stringError: boolean = false): boolean | string {
    const errors = []
    if (!this.verifySignature()) {
      errors.push('Invalid Signature')
    }

    if (stringError === false) {
      return errors.length === 0
    } else {
      return errors.join(' ')
    }
  }

  /**
   * sign a request with a given private key
   * @param privateKey
   */

  sign(privateKey: Buffer) {
    const sig = ecsign(this.msg(), privateKey)

    Object.assign(this, sig)
  }
}
